pkg/reduce/readloop/readloop.go

```diff
@@ -115,87 +115,156 @@ func (rl *ReadLoop) Startup(ctx context.Context) error {
-						return false, nil
-					}
-					return true, nil
-				})
-
-				if timeoutErr != nil {
-					rl.log.Errorw("Not able to write the messages to all partitions. To avoid message loss, this message will not be acked", zap.String("msgOffSet", m.ReadOffset.String()))
-					return
-				}
+		}
+		// happy path
+		rl.log.Debugw("Successfully wrote the message", zap.String("msgOffSet", m.ReadOffset.String()), zap.String("partitionID", p.String()))
+		return true, nil
+	})
 
-			}
+	return err
+}
 
-			// Ack the message to ISB
-			attempt := 0
-			timeoutErr := wait.ExponentialBackoff(pbqWriteBackoff, func() (done bool, err error) {
-				rErr := m.ReadOffset.AckIt()
-				attempt += 1
-				if rErr != nil {
-					rl.log.Errorw("Failed to ack message", zap.String("msgOffSet", m.ReadOffset.String()), zap.Int("attempt", attempt), zap.Error(rErr))
+// ackMessages acks messages. Retries until it can succeed or ctx.Done() happens.
+func (rl *ReadLoop) ackMessages(ctx context.Context, messages []*isb.ReadMessage) error {
+	var err error
+	var ackBackoff = wait.Backoff{
+		Steps:    math.MaxInt,
+		Duration: 1 * time.Second,
+		Factor:   1.5,
+		Jitter:   0.1,
+	}
+	// Ack the message to ISB
+	for _, m := range messages {
+		err := wait.ExponentialBackoff(ackBackoff, func() (done bool, err error) {
+			rErr := m.ReadOffset.AckIt()
+			if rErr != nil {
+				rl.log.Errorw("Failed to ack message", zap.String("msgOffSet", m.ReadOffset.String()), zap.Error(rErr))
+				// no point retrying if ctx.Done has been invoked
+				select {
+				case <-ctx.Done():
+					// no point in retrying after we have been asked to stop.
+					return false, ctx.Err()
+				default:
+					// keep retrying
 					return false, nil
 				}
-				rl.log.Debugw("Successfully acked message", zap.String("msgOffSet", m.ReadOffset.String()))
-
-				return true, nil
-			})
-
-			if timeoutErr != nil {
-				rl.log.Errorw("Timed out while trying to ack a message.", zap.String("msgOffSet", m.ReadOffset.String()))
-				return
 			}
+			rl.log.Debugw("Successfully acked message", zap.String("msgOffSet", m.ReadOffset.String()))
 
-			// close any windows that need to be closed.
-			wm := processor.Watermark(m.Watermark)
-			closedWindows := rl.windower.RemoveWindows(time.Time(wm))
-			rl.log.Infow("closing windows", zap.Int("length", len(closedWindows)), zap.Time("watermark", time.Time(wm)))
-
-			for _, cw := range closedWindows {
-				partitions := cw.Partitions()
-				rl.closePartitions(partitions)
-				rl.log.Debugw("Closing Window", zap.Time("windowStart", cw.StartTime()), zap.Time("windowEnd", cw.EndTime()))
-			}
+			return true, nil
+		})
+		// no point trying the rest of the message, most likely that will also fail
+		if err != nil {
+			rl.log.Errorw("Failed to ack message, exiting the loop", zap.Error(err))
+			break
```

Comment:
What if the writes to the PQB were successful and while acking those messages you received a ctx.done? Before you return you should ack all the messages in that batch right?

Reply:
That should be fine, this `break` is to get out of the loop only if the `Ack` is failing. Otherwise, it will run to completion.

Reply:
okay